README.md

# curatedPCaData <img src="man/figures/hex.png" align="right" height="139" />

<!-- badges: start -->
<!-- badges: end -->

## Overview

`curatedPCaData` is a collection of publically available and annotated
data resources concerning prostate cancer.

## Installation

You can install `curatedPCaData` from GitHub with:


    # install.packages("devtools")
    devtools::install_github("Syksy/curatedPCaData")

Above may fail depending on the connection stability, as the package is
relatively large. In this case it’s best to install via the tarball. To
build the package tarball from a cloned git repo, run the following in
terminal / command prompt while in the root of the project:

    R CMD build curatedPCaData

It is possible to install the self-built tarball or download a premade
latest package with (see Releases-page on GitHub):

    R CMD INSTALL curatedPCaData_x.y.z.tar.gz

<!--- add BioConductor once up --->

## Usage

Simple example use of curated datasets and ’omics there-in:


    library(curatedPCaData)

    curatedPCaData::mae_tcga
    ## A MultiAssayExperiment object of 10 listed
    ##  experiments with user-defined names and respective classes.
    ##  Containing an ExperimentList class object of length 10:
    ##  [1] cna.gistic: matrix with 23151 rows and 404 columns
    ##  [2] gex.rsem.log: matrix with 19658 rows and 461 columns
    ##  [3] mut: RaggedExperiment with 15389 rows and 371 columns
    ##  [4] cibersort: matrix with 22 rows and 417 columns
    ##  [5] xcell: matrix with 39 rows and 417 columns
    ##  [6] epic: matrix with 8 rows and 417 columns
    ##  [7] quantiseq: matrix with 11 rows and 417 columns
    ##  [8] mcp: matrix with 11 rows and 417 columns
    ##  [9] scores: matrix with 4 rows and 417 columns
    ##  [10] estimate: data.frame with 4 rows and 417 columns
    ## Functionality:
    ##  experiments() - obtain the ExperimentList instance
    ##  colData() - the primary/phenotype DataFrame
    ##  sampleMap() - the sample coordination DataFrame
    ##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
    ##  *Format() - convert into a long or wide DataFrame
    ##  assays() - convert ExperimentList to a SimpleList of matrices
    ##  exportClass() - save data to flat files

    curatedPCaData::mae_tcga[["gex.rsem.log"]][1:4,1:4]
    ##           TCGA.G9.6348.01 TCGA.CH.5766.01 TCGA.EJ.A65G.01 TCGA.EJ.5527.01
    ## ARHGEF10L          8.8729          8.5581          9.2085          8.7699
    ## HIF3A              5.9049          4.9716          6.7795          5.5978
    ## RNF17              0.4008          0.7574          0.0000          2.5554
    ## RNF10             12.3538         12.2950         11.9701         11.7983

    curatedPCaData::mae_tcga[["cna.gistic"]][1:4,1:4]
    ##         TCGA.2A.A8VL.01 TCGA.2A.A8VO.01 TCGA.2A.A8VT.01 TCGA.2A.A8VV.01
    ## ACAP3                 0               0               0               0
    ## ACTRT2                0               0               0               0
    ## AGRN                  0               0               0               0
    ## ANKRD65               0               0               0               0

    MultiAssayExperiment::colData(curatedPCaData::mae_tcga)[1:3,1:5]
    ## DataFrame with 3 rows and 5 columns
    ##               study_name   patient_id     sample_name        alt_sample_name overall_survival_status
    ##              <character>  <character>     <character>            <character>               <integer>
    ## TCGA.2A.A8VL        TCGA TCGA.2A.A8VL TCGA.2A.A8VL.01 F9F392D3-E3C0-4CF2-A..                       0
    ## TCGA.2A.A8VO        TCGA TCGA.2A.A8VO TCGA.2A.A8VO.01 0BD35529-3416-42DD-A..                       0
    ## TCGA.2A.A8VT        TCGA TCGA.2A.A8VT TCGA.2A.A8VT.01 BFECF807-0658-417B-9..                       0

    curatedPCaData::mae_taylor
    ## A MultiAssayExperiment object of 11 listed
    ##  experiments with user-defined names and respective classes.
    ##  Containing an ExperimentList class object of length 11:
    ##  [1] cna.gistic: matrix with 17832 rows and 194 columns
    ##  [2] cna.logr: matrix with 18062 rows and 218 columns
    ##  [3] gex.rma: matrix with 17410 rows and 179 columns
    ##  [4] mut: RaggedExperiment with 90 rows and 43 columns
    ##  [5] cibersort: matrix with 22 rows and 179 columns
    ##  [6] xcell: matrix with 39 rows and 179 columns
    ##  [7] epic: matrix with 8 rows and 179 columns
    ##  [8] quantiseq: matrix with 11 rows and 179 columns
    ##  [9] mcp: matrix with 11 rows and 179 columns
    ##  [10] scores: matrix with 4 rows and 179 columns
    ##  [11] estimate: data.frame with 4 rows and 179 columns
    ## Functionality:
    ##  experiments() - obtain the ExperimentList instance
    ##  colData() - the primary/phenotype DataFrame
    ##  sampleMap() - the sample coordination DataFrame
    ##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
    ##  *Format() - convert into a long or wide DataFrame
    ##  assays() - convert ExperimentList to a SimpleList of matrices
    ##  exportClass() - save data to flat files

    curatedPCaData::mae_sun
    ## A MultiAssayExperiment object of 8 listed
    ##  experiments with user-defined names and respective classes.
    ##  Containing an ExperimentList class object of length 8:
    ##  [1] gex.rma: matrix with 12784 rows and 79 columns
    ##  [2] cibersort: matrix with 22 rows and 79 columns
    ##  [3] xcell: matrix with 39 rows and 79 columns
    ##  [4] epic: matrix with 8 rows and 79 columns
    ##  [5] quantiseq: matrix with 11 rows and 79 columns
    ##  [6] mcp: matrix with 11 rows and 79 columns
    ##  [7] scores: matrix with 4 rows and 79 columns
    ##  [8] estimate: data.frame with 4 rows and 79 columns
    ## Functionality:
    ##  experiments() - obtain the ExperimentList instance
    ##  colData() - the primary/phenotype DataFrame
    ##  sampleMap() - the sample coordination DataFrame
    ##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
    ##  *Format() - convert into a long or wide DataFrame
    ##  assays() - convert ExperimentList to a SimpleList of matrices
    ##  exportClass() - save data to flat files

## R Shiny

A web interface built with R Shiny for `curatedPCaData` can be launched
via:

    curatedPCaData::shiny()

## Known issues

Large R-packages directly installed using `devtools::install_github`
sometimes result in error:

    Error in utils::download.file( ...
        download from ' ... ' failed.

There are few options to fix this:

### Adjusting download options

For Windows users, it’s known that setting

    options(download.file.method = "wininet")

may fix this GitHub direct download/install issue. Alternatively,
especially in other OSes, the following download method may fix this
issue:

    options(download.file.method = "libcurl")

### Direct cloning, building, or downloading and installing the package tarball

A more hands-on approach is to use Git to clone the package and then
build and install it using R tools. In your terminal or command prompt,
go to a suitable root directory for git repositories:

    git clone https://github.com/Syksy/curatedPCaData.git
    R CMD build curatedPCaData

alternatively, to speed up package tarball building, add parameter
`--no-build-vignettes`. This will produce a file
`curatedPCaData_x.y.z.tar.gz` in your active directory, where `x.y.z`
correspond to the current package version.

Alternatively, if one does not wish to clone the git and build the
tarball, it’s also possible to directly download latest release tarball
from
<a href="https://github.com/Syksy/curatedPCaData/releases" class="uri">https://github.com/Syksy/curatedPCaData/releases</a>

After this, the `curatedPCaData`R-package tarball can be installed
using:

    R CMD INSTALL curatedPCaData_x.y.z.tar.gz

Please note that some dependencies (such as the packages
`MultiAssayExperiment` and `S4Vectors`) may produce an error during
installation if they are not found for R. In this case these
dependencies need to be installed from their respective repositories
such as CRAN or BioConductor.


NEWS
curatedPCaData v0.9.33 (Release date: 2022-09-27)
==============

Changes:

* Revised vignettes
* Revised parameters for running immunedeconv (esp. quanTIseq)
* Adding ESTIMATE from immunedeconv >= v2.1.0
* Subset to colData sample_name-column when adding new MAE slots (i.e. new assay/platform types)

curatedPCaData v0.9.32 (Release date: 2022-09-15)
==============

Changes:

* Clinical data revised for TCGA (xenabrowser-version) with a larger N subset appended by cBio's additional fields when available
* Brief edits to other datasets' clinical metadata as well
* Revised vignettes and new material

curatedPCaData v0.9.31 (Release date: 2022-09-08)
==============

Changes:

* CNA GISTIC for Hieronymus et al. replacing the old log-ratios MAE slot
* Fixes in landscapes.Rmd

curatedPCaData v0.9.30 (Release date: 2022-09-06)
==============

Changes:

* CNA landscape plots (original code thanks to Mike)

curatedPCaData v0.9.29 (Release date: 2022-09-05)
==============

Changes:

* cBioPortal processing pipeline revised
* MAEs and corresponding raw data revised
* CIBERSORTx rerun

curatedPCaData v0.9.28 (Release date: 2022-07-25)
==============

Changes:

* Reprocessed data

curatedPCaData v0.9.27 (Release date: 2022-07-18)
==============

Changes:

* Lifted computational burden from a vignette to pre-computed workspace
* Fixed class check comparison WARNINGS from BiocCheck with methods::is 

curatedPCaData v0.9.26 (Release date: 2022-07-15)
==============

Changes:

* CNA data matrices filtered down to protein coding (i.e. exclude miRNAs and such)

curatedPCaData v0.9.25 (Release date: 2022-07-14)
==============

Changes:

* biocViews / BiocType expanded, various BiocCheck fixes
* Corrected internal function/data calls
* Various misc fixes so that R CMD check is clean (except size NOTE)
* Testing Bioconductor instructions for LazyData set to false, replaced by use of 'loadPCa()'-function by the user

curatedPCaData v0.9.24 (Release date: 2022-07-13)
==============

Changes:

* R CMD check feedback fixes
* BiocCheck feedback fixes
* biocViews in DESCRIPTION (ExperimentData)

curatedPCaData v0.9.23 (Release date: 2022-07-08)
==============

Changes:

* Median shift bug in risk score calculations fixed
* Fixing R CMD check Notes

curatedPCaData v0.9.21 (Release date: 2022-06-23)
==============

Changes:

* Updated NAMESPACE and Rds

curatedPCaData v0.9.20 (Release date: 2022-06-22)
==============

Changes:

* LazyDataCompression added
* Overhaul of dependencies

curatedPCaData v0.9.19 (Release date: 2022-06-21)
==============

Changes:

* Fixes to vignettes
* Additional allowed values in template_prad's 'grade_group' for cases when info such as 4+3 or 3+4 is not available but instead just 7 was available in original source
* Additional necessary content, such as this NEWS file and inst/CITATION

curatedPCaData v0.9.18 (Release date: 2022-06-17)
==============

Changes:

* Moving through checklists for BioconductoR compatible experiment data package structure


R/pipelines.R
###
#
# Sun, et al. portion
#
###

#' Generate gene expression data for Sun et al. 
#' 
#' @param file_directory Character string indicating directory for downloading files. 
#' Files may be large, so please check that there is sufficient free space. 
#' If NULL then files are downloaded into current directory.
#' @param cleanup Logical. Remove tarballs and other files from working directory.
#' @param ... Additional arguments. 
#'
#' @return Gene expression object of a particular type 
#'
#' @export
Generate_GEX_Sun <- function(
	file_directory, 
	cleanup = FALSE, 
	collapseFUN = function(z) { apply(z, MARGIN=2, FUN=median) }, # Function to collapse probe(s) or select a probe, e.g. mean, median, or function that picks a probe with high variance
	...
){
	if(!missing(file_directory)) setwd(file_directory) # exchange setwd with here::here()
	# Supplementary files include the raw CEL files
	supfiles <- GEOquery::getGEOSuppFiles('GSE25136')
	# Download size: 269.5 MB
	# Open the tarball(s)
	utils::untar(tarfile=rownames(supfiles))
	# Make sure to function in a working directory where the are no other tarballs present
	supfiles2 <- list.files()
	supfiles2 <- supfiles2[grep(".gz", supfiles2)]
	#invisible(lapply(supfiles2, FUN=gunzip))
	# Read Affymetrix MA
	Sun <- affy::ReadAffy()
	colnames(affy::exprs(Sun)) <- gsub(".gz|.CEL", "", colnames(Sun))
	# Careful not to mask 'rma' from 'affy' by the 'rma' from 'oligo'
	GEX_Sun <- affy::rma(Sun)
	# Removing .CEL and packaging names from the GEO-compatible sample names
	colnames(GEX_Sun) <- gsub(".CEL.gz", "", colnames(affy::exprs(GEX_Sun)))
	keys <- AnnotationDbi::mappedkeys(hgu133a.db::hgu133aGENENAME)
	nam <- names(as.character(hgu133a.db::hgu133aALIAS2PROBE)[match(rownames(GEX_Sun), as.character(hgu133a.db::hgu133aALIAS2PROBE))])
	nam[is.na(nam)] <- "NA"
	# Collapse probes
	GEX_Sun <- do.call("rbind", by(as.matrix(affy::exprs(GEX_Sun)), INDICES=nam, FUN=collapseFUN))
	# Remove downloaded files
	if(cleanup){
		# First GEO download
		file.remove(rownames(supfiles))
		# Tarballs
		file.remove(supfiles2)
	}
	# Return numeric matrix
	as.matrix(GEX_Sun)
}

#' Generate MultiAssaYExperiment-object for Sun, et al.
#'
#' TODO
#'
#' @export
Generate_MAE_Sun <- function(
	pdata = "../data-raw/GSE25136_curated_pdata.txt",
	image.file,	# If user wishes to save an .RData object of the MAE-object, a file name can be specified
	cleanup,	# Whether intermediate files ought to be cleaned
	...
){
	# Generate gene expression matrix from GEO
	GEX_Sun <- Generate_GEX_Sun(cleanup = cleanup)
	# Start constructing colData
	tmpColDat <- S4Vectors::DataFrame(
		read.table(pdata, # Curated data dictionary file
		sep="\t", header=TRUE)
	)
	rownames(tmpColDat) <- tmpColDat$sample_name
	# Generate a MAE-object for Sun GEX
	MAE_Sun <- MultiAssayExperiment::MultiAssayExperiment(
		experiments = MultiAssayExperiment::ExperimentList(
			"GEX" = GEX_Sun
		),
		colData = tmpColDat,
		sampleMap <- S4Vectors::DataFrame(
			rbind(
				data.frame(assay = "GEX", primary = tmpColDat$patient_id, colname = tmpColDat$sample_name)
			)
		)
	)
	# Optionally save an external R workspace file with the MAE object
	if(!missing(image.file)) save(MAE_Sun, file=image.file)
	# Return MAE object
	MAE_Sun
}


####
#
# Taylor, et al. portion
#
###

#' Gene expression data from Taylor et al.
#'
#' TODO
#'
#' GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE21032
#' Available data types; Affymetrix Exon arrays, 2 separate types
#'
#' @export
Generate_GEX_Taylor <- function(
	file_directory,
	cleanup = FALSE,
	...
){
	if(!missing(file_directory)) setwd(file_directory) # exchange setwd with here::here()
	# Fetch supplementary files from Taylor et al. from GEO
	# Be wary as the tarball already is 24.2 Gb
	supfiles <- GEOquery::getGEOSuppFiles('GSE21032')
	# Step in the internal directory created by GEOquery
	setwd("GSE21032")
	# Open the tarball
	utils::untar(tarfile=rownames(supfiles))
	# First download CEL files from GEO
	CELs <- oligo::read.celfiles(affy::list.celfiles())	
	#> CELs
	#ExonFeatureSet (storageMode: lockedEnvironment)
	#assayData: 6553600 features, 370 samples 
	#  element names: exprs 
	#protocolData
	#  rowNames: GSM526134_YX_Exon1_PCA0001.CEL GSM526135_YX_Exon1_PCA0002.CEL ... GSM528049_YX_Exon1_PAN0174.CEL (370
	#    total)
	#  varLabels: exprs dates
	#  varMetadata: labelDescription channel
	#phenoData
	#  rowNames: GSM526134_YX_Exon1_PCA0001.CEL GSM526135_YX_Exon1_PCA0002.CEL ... GSM528049_YX_Exon1_PAN0174.CEL (370
	#    total)
	#  varLabels: index
	#  varMetadata: labelDescription channel
	#featureData: none
	#experimentData: use 'experimentData(object)'
	#Annotation: pd.huex.1.0.st.v2
	RMAs <- oligo::rma(CELs)
	featureData(RMAs) <- oligo::getNetAffx(RMAs, "transcript")
	sampleNames(RMAs) <- unlist(lapply(strsplit(affy::list.celfiles(), "_"), FUN=function(z) z[[1]])) # GSM######-type names from GEO
	nameMapTaylor <- data.frame(
		GSM = unlist(lapply(strsplit(affy::list.celfiles(), "_"), FUN=function(z) z[[1]])), # GEO, GSM-names
		ID = gsub(".CEL|.gz", "", unlist(lapply(strsplit(affy::list.celfiles(), "_"), FUN=function(z) z[[4]]))) # mapping to PCA0001, PCA0002, ...
	)
	# 
	GEX_Taylor <- RMAs
	# Proper naming
	nam <- fData(GEX_Taylor)[,8]
	nam2 <- unlist(lapply(nam, FUN=function(z) { strsplit(z, " // ")[[1]][2] }))
	nam2[is.na(nam2)] <- "NA"
	rownames(GEX_Taylor) <- make.unique(nam2)
	## TODO:
	if(cleanup){
		# Remove tarballs and other potential intermediary files
	}
	
	# Return constructed gene expression matrix
	GEX_Taylor
}

#' Copy number alterations from Taylor et al.
#'
#' TODO
#' GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE21035
#' Agilent aCGH
#' Platform in GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL4091
#' Agilent-014693 Human Genome CGH Microarray 244A
#' Ref: Taylor BS, Schultz N, Hieronymus H, Gopalan A et al. Integrative genomic profiling of human prostate cancer. Cancer Cell 2010 Jul 13;18(1):11-22. 
#' 
#' @export
Generate_CNA_Taylor <- function(
	file_directory, 
	cleanup = FALSE,
	...
){
	if(!missing(file_directory)) setwd(file_directory) # exchange setwd with here::here()
	# Fetch supplementary files from Taylor et al. CNA from GEO
	# Be wary as the tarball already is 14.9 Gb
	supfiles <- GEOquery::getGEOSuppFiles('GSE21035')
	# Step in the internal directory created by GEOquery
	setwd("GSE21035")
	# Open the tarball
	utils::untar(tarfile=rownames(supfiles))
	# Read the raw Agilent CGH data
	# Name samples according to the filename
	# Agilent-014693 Human Genome CGH Microarray 244A (Feature number version)
	# rCGH says for readAgilent-function: "Agilent from 44 to 400K are supported."
	TaylorCGH <- lapply(base::list.files(), FUN=function(z) { 
		try({
			cat("\n\nProcessing: ",z,"\n\n"); 
			rCGH::readAgilent(z, genome="hg38", sampleName=gsub(".txt", "", z)) 
		})
	})
	#> list.files()[which(unlist(lapply(TaylorCGH, FUN=class))=="try-error")]
	#[1] "GSM525755.txt" "GSM525763.txt"
	# Some files appear broken; missing columns?
	
	# Signal adjustments
	TaylorCGH <- lapply(TaylorCGH, FUN=function(z){
		try({
			rCGH::adjustSignal(z) 
		})
	})
	# Segmentation
	TaylorCGH <- lapply(TaylorCGH, FUN=function(z){
		try({
			rCGH::segmentCGH(z) 
		})
	})
	# EM-algorithm normalization
	TaylorCGH <- lapply(TaylorCGH, FUN=function(z){
		try({
			rCGH::EMnormalize(z) 
		})
	})
	# For the old pipeline with lacking sampleName but included fileName, run:
	CNA_Taylor <- lapply(CNA_Taylor, FUN=.renameSampleName)
	### TODO:
	
	# cleanup
	
	## Export a compiled GISTIC 2.0 -compatible file from the segmented samples
	##exportGISTIC(CNA_Taylor, file="inputGISTIC_Taylor.tsv")
	CNA_Taylor
	
}	

#' Generate MultiAssayExperiment-object for Taylor, et al.
#'
#' TODO
#'
#' @export
Generate_MAE_Taylor <- function(
	pdata = "../data-raw/prad_mskcc_curated_pdata.txt",
	image.file,	# If user wishes to save an .RData object of the MAE-object, a file name can be specified
	cleanup=FALSE,	# Whether intermediate files ought to be cleaned
	...
){
	# Gene expression from GEO
	GEX_Taylor <- Generate_GEX_Taylor(cleanup=cleanup)
	# Copy number alterations from GEO
	CNA_Taylor <- Generate_CNA_Taylor(cleanup=cleanup)
	
	# Running Taylor MAE:
	
	# Start constructing colData
	tmpColDat <- S4Vectors::DataFrame(
		#read.table(pdata, # Curated data dictionary file
		read.table("D:/Gits/curatedPCaData/data-raw/prad_mskcc_curated_pdata.txt", # Curated data dictionary file
		sep="\t", header=TRUE)
	)
	# Set row names
	rownames(tmpColDat) <- tmpColDat$patient_id

	# Generate a MAE-object for TCGA 'omics
	MAE_Taylor <- MultiAssayExperiment::MultiAssayExperiment(
		experiments = MultiAssayExperiment::ExperimentList(
			"GEX" = t(GEX_Taylor)#,
			#"CNA" = t(CNA_Taylor)
		),
		colData = tmpColDat,
		sampleMap <- S4Vectors::DataFrame(
			rbind(
				data.frame(assay = "GEX", primary = tmpColDat$patient_id, colname = tmpColDat$sample_name)#,
				#data.frame(assay = "CNA", primary = tmpColDat$patient_id, colname = tmpColDat$sample_name)
			)
		)
	)	
	
	# Optionally save an external R workspace file with the MAE object
	if(!missing(image.file)) save(MAE_Taylor, file=image.file)
	# Return MAE-object from Taylor et al.
	MAE_Taylor
}

####
#
# TCGA portion
#
###

#' Generate a MultiAssayExperiment-object of TCGA GEX + CNA
Generate_MAE_TCGA <- function(
	pdata = "../data-raw/prad_tcga_curated_pdata.txt",	# Location of the curated pdata txt-file
	image.file,	# If user wishes to save an .RData object of the MAE-object, a file name can be specified
	# clean,	# No clean-up required for TCGA as no intermediate files are generated
	...
){
	# Running TCGA GEX:
	GEX_TCGA <- Generate_cBioPortal(genes = genes$hgnc, geneticProfiles="prad_tcga_rna_seq_v2_mrna", caseList="prad_tcga_sequenced")

	# Running TCGA CNA:
	CNA_TCGA <- Generate_cBioPortal(genes = genes$hgnc, geneticProfiles="prad_tcga_gistic", caseList="prad_tcga_sequenced")

	# Running TCGA MAE:
	
	# Start constructing colData
	tmpColDat <- S4Vectors::DataFrame(
		read.table(pdata, # Curated data dictionary file
		sep="\t", header=TRUE)
	)
	# Sanitize '-'-symbols in TCGA names as R tends to replace mathematical operators or whitespace with '.'s
	tmpColDat$sample_name <- gsub("-", ".", tmpColDat$sample_name, fixed=TRUE)
	# Set row names
	rownames(tmpColDat) <- tmpColDat$patient_id

	# Generate a MAE-object for TCGA 'omics
	MAE_TCGA <- MultiAssayExperiment::MultiAssayExperiment(
		experiments = MultiAssayExperiment::ExperimentList(
			"GEX" = t(GEX_TCGA),
			"CNA" = t(CNA_TCGA)
		),
		colData = tmpColDat,
		sampleMap <- S4Vectors::DataFrame(
			rbind(
				data.frame(assay = "GEX", primary = tmpColDat$patient_id, colname = tmpColDat$sample_name),
				data.frame(assay = "CNA", primary = tmpColDat$patient_id, colname = tmpColDat$sample_name)
			)
		)
	)
	# Optionally save an external R workspace file with the MAE object
	if(!missing(image.file)) save(MAE_TCGA, file=image.file)
	# Return MAE object
	MAE_TCGA
}

#' Generic function for pulling data from cBioPortal; useful for e.g. TCGA pulling on multiple omics layers
Generate_cBioPortal <- function(
	genes, # List of gene symbols to iterate over
	geneticProfiles="prad_tcga_rna_seq_v2_mrna", # for cgdsr calls, platform and dataset specific string
	caseList="prad_tcga_sequenced", # for cgdsr calls, platform and dataset specific string
	...
){
	#http://www.cbioportal.org/study?id=prad_tcga#summary
	mycgds <- cgdsr::CGDS("http://www.cbioportal.org/")
	# Use the wrapper function to iterative calls for split gene lists
	dat <- getProfileDataWrapper(
		x=mycgds, # cgdsr object
		#genes=curatedPCaData:::.getGeneNames()$hgnc, # All unique gene symbols
		genes=genes, # All unique gene symbols
		geneticProfiles=geneticProfiles, # Omics profile
		caseList=caseList # Case list
	)
	# TODO: Filter out low quality samples based on list provided by Travis
	# Return omics matrix
	dat
}

###
#
# ICGC portion
#
###

## Multiple specific data pulls from the ICGC depo

#' PRAD-CA (PRAD-CA Prostate Adenocarcinoma - CA)
Generate_ICGC_CA <- function(
	file_directory, 
	# Relevant available files for the Canadian ICGC dataset
	filelist = c(
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/copy_number_somatic_mutation.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/donor.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/donor_exposure.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/donor_family.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/donor_therapy.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/exp_array.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/meth_array.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/sample.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/simple_somatic_mutation.open.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/specimen.PRAD-CA.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-CA/structural_somatic_mutation.PRAD-CA.tsv.gz"
	)
){
	if(!missing(file_directory)) setwd(file_directory) # exchange setwd with here::here()
	ICGC.PRAD.CA <- lapply(filelist, FUN=.icgcDownload)
	ICGC.PRAD.CA
}

## PRAD-CN (PRAD-CN Prostate Cancer - CN)
## -> OMIT?

#' PRAD-FR (PRAD-FR Prostate Cancer - Adenocarcinoma - FR)
Generate_ICGC_FR <- function(
	file_directory, 
	# Relevant available files for the Canadian ICGC dataset
	filelist = c(
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/copy_number_somatic_mutation.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/donor.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/donor_family.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/donor_surgery.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/exp_array.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/exp_seq.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/sample.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/simple_somatic_mutation.open.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/specimen.PRAD-FR.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-FR/structural_somatic_mutation.PRAD-FR.tsv.gz"
	)
){
	if(!missing(file_directory)) setwd(file_directory) # exchange setwd with here::here()
	ICGC.PRAD.FR <- lapply(filelist, FUN=.icgcDownload)
	ICGC.PRAD.FR
}


#' PRAD-UK (PRAD-UK Prostate Adenocarcinoma - UK)
Generate_ICGC_UK <- function(
	file_directory, 
	# Relevant available files for the Canadian ICGC dataset
	filelist = c(
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/copy_number_somatic_mutation.PRAD-UK.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/donor.PRAD-UK.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/donor_exposure.PRAD-UK.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/donor_family.PRAD-UK.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/donor_therapy.PRAD-UK.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/sample.PRAD-UK.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/simple_somatic_mutation.open.PRAD-UK.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/specimen.PRAD-UK.tsv.gz",
		"https://dcc.icgc.org/api/v1/download?fn=/release_27/Projects/PRAD-UK/structural_somatic_mutation.PRAD-UK.tsv.gz"
	)
){
	if(!missing(file_directory)) setwd(file_directory) # exchange setwd with here::here()
	ICGC.PRAD.UK <- lapply(filelist, FUN=.icgcDownload)
	ICGC.PRAD.UK
}

## PRAD-US (PRAD-US Prostate Adenocarcinoma - TCGA, US)
## -> OMIT?

####
#
# Supporting variables
#
####

# Extract whole list of desired gene names in various formats
genes <- .getGeneNames()

####
#
# Sun, et al.
# GEX
#
####

# Running Sun, et al.:
MAE_Sun <- Generate_MAE_Sun()

####
#
# TCGA 
# GEX + CNA
#
####

## Get the summary lists for 'omics and case profiles (gives names for omics 'geneticProfiles' and 'caseList')
# cgdsr::getCaseLists(mycgds,"prad_tcga")
# cgdsr::getGeneticProfiles(mycgds,"prad_tcga")

# Note from Jordan & Christelle 
# N=333 in TCGA provisional, Cell 2015

MAE_TCGA <- Generate_MAE_TCGA()


####
#
# Taylor, et al.
# GEX + CNA
#
####

